rt_segment_speeds/scripts/B2_avg_speeds_by_segment.py

```diff
@@ -108,23 +137,29 @@ def speeds_with_segment_geom(
         peak.assign(time_of_day = "peak")
     ], axis=0)
     
+    # start with segments with geom (scheduled, we have way too many)
+    # merge against unique_segments (these are present in RT...inner join)...we have geom 
+    missing = (pd.merge(unique_segs_with_geo, 
+                        stats, 
+                        on = ['shape_array_key', 'stop_sequence'], 
+                        how = "left", 
+                        indicator = True))
     
-    # Merge in segment geometry
-    segments = helpers.import_segments(
-        SEGMENT_GCS,
-        f"{SEGMENT_FILE}_{analysis_date}",
-        columns = SEGMENT_IDENTIFIER_COLS + [
-            "schedule_gtfs_dataset_key", 
-            "stop_id",
-            "loop_or_inlining",
-            "geometry", 
-            "district", "district_name"
-        ]
-    ).to_crs(geography_utils.WGS84)
+    # Grab left only results, which did not show up in the stats df
+    missing = missing.loc[missing._merge == "left_only"].reset_index(drop = True)
+    missing = missing.drop(columns = ['_merge'])
+    
+    # Concat & clean.
+    stats2 = pd.concat([missing, stats])
+    stats2 = stats2.fillna(stats2.dtypes.replace({'float64': 0.0, 'object': 'None'}))
```

Comment:
What are the `stats2 = pd.concat / fillna` steps doing here? Instead, if in line 142, you do 2 left merges of segment geom with `all_day` and then `peak`, that would get you the same result

Reply:
That was a way to bring back the unique segments that were dropped and concat them with the stats dataframe. The fillna is to replace any of the segments that were dropped with 0's instead of nulls and none instead of peak/all day.

So I should do something like this? 
`
stats = pd.concat([
        all_day.assign(time_of_day = "all_day"),
        peak.assign(time_of_day = "peak")
    ], axis=0)

segments = segments.to_crs(geography_utils.WGS84)

gdf = pd.merge(
        segments,
        stats,
        on = SEGMENT_IDENTIFIER_COLS,
        how = "left"
    )
`

Reply:
Yes, I prefer left joins. Filling with zeros might mean that `df[df.speed_mph==0]` grabs a completely different set of rows from `speeds_{date}` and `avg_speeds_{date}`. If there are zeroes in speeds, then somehow removing them and averaging them still results in a row with 0's...that could be confusing.